addon/classes/ChangesetObjectProxyHandler.ts

```diff
@@ -0,0 +1,394 @@
+        // we have a pending change
+        // return it
+        // we know it's not undefined so we can safely cast it
+        var change = changes.get(localKey);
+        var value = getTrackedValue(change as TrackedStorage<any>);
+        if (value !== ObjectReplaced) {
+          return value
+        }
+      }
+    }
+    // drop back to the internal object property
+    // or a proxy of it if it's an object
+
+    // @ts-ignore
+    var value = this.__source[localKey];
+    if (requiresProxying(value)) {
+      // we know that this key has not already been proxied
+      var proxy = this.addProxy(localKey);
+      if (subkey) {
+        return proxy[subkey];
+      }
+      return proxy;
+    }
+    return value;
+  }
+
+  private addProxy(key: string, value?: {}): any {
+    // get sends just the key
+    // set sends the key and the new value
+    if (value === undefined) {
+      // use the original
+      value = this.__source[key];
+    }
+    if (value === undefined) {
+      // missing on original but added in the changeset
+      value = {};
+    }
+    var proxy = new Proxy(value, handlerFor(value));
+    this.__nestedProxies.set(key, proxy);
+    return proxy;
+  }
+
+  @action
+  public setValue(key: string, value: any): boolean {
+    switch (key) {
+      case 'change':
+      case 'changes':
+      case 'data':
+      case 'isChangeset':
+      case 'isDirty':
+      case 'isPristine':
+      case 'isValid':
+      case 'pendingData':
+        throw `changeset.${key} is readonly`;
+    }
+    // nested keys are separated by dots
+    var [localKey, subkey] = splitKey(key as string);
+
+    if (subkey) {
+      // pass the change down to a nested level
+      var proxy = this.__nestedProxies.get(localKey);
+      if (!proxy) {
+        // no existing proxy
+        // so they're trying to set deep into an object that isn't yet proxied
+        // wrap the existing object or create an empty one
+        proxy = this.addProxy(localKey);
+      }
+      return proxy.set(subkey, value);
+    } else {
+      // this is a change at our level
+      // check the changeset key filter
+      if (this.isKeyFilteredOut(localKey)) {
+        return false;
+      }
+      if (requiresProxying(value)) {
+        var proxy = this.addProxy(localKey, value);
+        // remove a tracked value if there was one with the same key
+        this.markChange(localKey, ObjectReplaced);
+      } else {
+        // the value is a local property
+        this.markChange(localKey, value);
+        // remove a proxy if there was one with the same key
+        if (this.__nestedProxies.has(key)) {
+          this.__nestedProxies.delete(key);
+        }
+      }
+      return true;
+    }
+  }
+
+  private markChange(localKey: string, value: any) {
+    // have to use get() here because source might be an EmberProxy
+    const oldValue = get(this.__source, localKey);
+    const unchanged = isUnchanged(value, oldValue);
+    let changes = this.__changes;
```

Comment:
Wait, does this PR remove and re-implement validated-changeset?

Sorry if obvious